### ISSUE TITLE
Investigate why converting a safe document leads to size inflation

### ISSUE THREAD

**Issue by apyrgio** [freedomofpress/dangerzone#239]
When we attempt to reconvert a safe document with Dangerzone, we've seen that it leads to size inflation in the resulting file. After roughly 4 re-conversions, Dangerzone times out.

We've noticed that the image dimensions of the scanned document get bigger and bigger each time, so our assumption is that some step in the process is not registering correctly the width and height of the original image. We need to find the faulty step, and fix it.

**Comment by almet** [comment]
I've been hit by this recently when sanitizing PDFs with images in it. I believe it applies the same way when just using images.

Right now, Dangerzone stores pictures as raw pixels with lossless compression (deflate), no matter what's on the page originally, and no matter the original compression algorithm.

That seem to be generally fine for text, but terrible for photos and scans where the output ends up 3 to 4 times bigger.

---

So, what's the solution?

Ideally, we'd want to use a different compression algorithm for text-based documents and photos. Using JPEG with q75 compression is probably what we want for pictures, but not for text-only documents, where that would lead to a size inflation.

I've made an experimentation with a simple heuristic that detects the "dominant color percentage", with this code:

```python
from collections import Counter

# Get one pixel every 97 to avoid getting pixels from the same columns.
SAMPLE = 97

def dominant_color_percentage(rgb_data: bytes) -> float:
    """Returns the percentage of (sampled) pixels that share the most common color."""

    # One pixel = 3 consecutive bytes. Get them with `i : i+3` slices.
    pixels = [
        rgb_data[i : i + 3]
        for i in range(0, len(rgb_data) - 3, 3 * SAMPLE)
    ]

    if not pixels:
        return 0.0

    (_, count), = Counter(pixels).most_common(1)

    return count / len(pixels)
```

Here are the results with a 80% gate:

| Page                               | Dominant color | deflate | JPEG q75 | Winner  | Heuristic ok |
|------------------------------------|----------------|---------|----------|---------|--------------|
| sample-pdf.pdf p1 (text)           | 92.5%          | 162K    | 174K     | deflate | yes          |
| sample-pdf.pdf p2 (text)           | 91.3%          | 200K    | 238K     | deflate | yes          |
| sample-pdf.pdf p3 (text)           | 91.6%          | 198K    | 233K     | deflate | yes          |
| sample-pdf.pdf p4 (text)           | 98.9%          | 36K     | 52K      | deflate | yes          |
| [dangerzone-stacktower-safe.pdf](https://github.com/user-attachments/files/30653022/dangerzone-stacktower-safe.pdf)                 | 25.7%          | 392K    | 203K     | JPEG    | yes          |
| DZ screenshot1.png                 | 43.5%          | 143K    | 78K      | JPEG    | yes          |
| [Photo of a doc](https://github.com/user-attachments/assets/76f6b797-9f65-4d85-a5ef-7b89f0fd91e0) | 1.0%           | 1765K   | 123K     | JPEG    | yes          |

@apyrgio what's your feeling on using this heuristic and using a JPEG q75 compression algorithm by default when the ratio is > 0.80?

**Comment by apyrgio** [comment]
Oooh, that's very interesting! Looking into this subject, a few things come to mind:

1. The sampling and dominant color proposal look interesting, and safe. What I'm a bit hesitant about is adding a JPEG encoding step :-/. We've seen several issues with such encoders/decoders even to this day, so I think it's a more apt fit for our rewrite in Rust.
2. The Internet Archive handles this a bit differently using [Mixed Raster Content](https://github.com/internetarchive/archive-pdf-tools). Same concerns apply, and this time conversion may take much more time. I'm not proposing it seriously, but I figured it's interesting to have it in mind.
3. If our sandbox starts printing out media types (e.g., document, image, audio, video), we may also use that as a heuristic as well. Of course, a document with embedded images may not compress well, and that's where the sampling can help.

**Comment by almet** [comment]
> I'm a bit hesitant about is adding a JPEG encoding step :-/.

you got me curious :-) Can you expand on that?

**Comment by yveszoundi** [comment]
Maybe a viable approach is to maintain a maximum size threshold while extracting "documents" (MuPDF docs) into images.

"When target width is greater than max width Or target height is greater than max height, Then scale doc before rendering."

**Comment by almet** [comment]
@yveszoundi thanks for hopping in. I don't think this is related to the height or width of the produced documents, unfortunately. To me the size inflation is related to the compression algorithm (we decompress and then recompress with a sub-optimal algo).

We've discussed live with @apyrgio and his concerns are with memory safety implementations of JPEG, and so we might want to use a rust library for this.

https://lib.rs/crates/jpeg-encoder seem to be one, and we could either embed it in dangerzone and have python bindings to it via maturin. We could propose this python package for others to use, and potentially create a Debian and Fedora package so it's then packaged upstream, and we would vendorize it for now in our packages.

Or… we could wait for the rust implementation to be ready in order to solve this problem.

**Comment by apyrgio** [comment]
Regarding what PyMuPDF can offer us right now, here's what we can use: https://pymupdf.readthedocs.io/en/latest/document.html#Document.rewrite_images

**Comment by almet** [comment]
The muPDF route seem to come with the same issues as doing the compression in python: it's not memory safe, and I think that's what you meant at first when saying that you wanted to avoid JPEG encoding, right?

**Comment by apyrgio** [comment]
Correct, I'm mentioning it for completeness, in the off-chance they are using a memory-safe codec somehow. Haven't checked yet though.
